File: packages/python/onnx9000-onnx2gguf/src/onnx9000/onnx2gguf/compiler.py
from .naming import rename_tensor
from .tokenizer import extract_tokenizer_metadata
import re
from typing import BinaryIO, Any, Optional
from onnx9000.core.ir import Graph, Node, Tensor
from onnx9000.core.dtypes import DType
from .builder import GGUFWriter, GGUFValueType, GGUFTensorType


def get_gguf_type(dtype: int) -> GGUFTensorType:
    if dtype == DType.FLOAT32.value:
        return GGUFTensorType.F32
    if dtype == DType.FLOAT16.value:
        return GGUFTensorType.F16
    return GGUFTensorType.F32


def infer_architecture(graph: Graph) -> str:
    # Very basic heuristic for Phase 2, to be expanded in Phase 3/4
    # Check tensor names or node types
    text = str(graph.name) + str(graph.tensors.keys()) + str([n.op_type for n in graph.nodes])
    if "llama" in text.lower() or "Llama" in text:
        return "llama"
    return "unknown"


def sanitize_doc_string(doc: str) -> str:
    if not doc:
        return ""
    # Strip non-utf8 or weird controls if needed
    return doc.strip()


def compile_gguf(graph: Graph, out_f: BinaryIO, kv_overrides: Optional[dict[str, Any]] = None):
    writer = GGUFWriter(out_f)
    kv_overrides = kv_overrides or {}

    arch = infer_architecture(graph)

    # Phase 2 defaults
    general_kvs = {
        "general.architecture": arch,
        "general.name": graph.name if graph.name else "model",
        "general.author": getattr(graph, "producer_name", "onnx9000") or "onnx9000",
        "general.version": getattr(graph, "model_version", 1) or 1,
        "general.quantization_version": 2,
        "general.alignment": 32,
        "general.file_type": "mostly_f32",
    }

    doc_str = getattr(graph, "doc_string", "")
    if doc_str:
        general_kvs["general.description"] = sanitize_doc_string(doc_str)

    for k, v in kv_overrides.items():
        if k.startswith("general."):
            general_kvs[k] = v

    for k, v in general_kvs.items():
        if isinstance(v, bool):
            writer.add_bool(k, v)
        elif isinstance(v, str):
            writer.add_string(k, v)
        elif isinstance(v, int):
            writer.add_uint32(k, v)
        elif isinstance(v, float):
            writer.add_float32(k, v)

    # Llama specifics (if it is llama)
    if general_kvs["general.architecture"] == "llama":
        writer.add_uint32("llama.context_length", 2048)
        writer.add_uint32("llama.embedding_length", 4096)
        writer.add_uint32("llama.block_count", 32)
        writer.add_uint32("llama.feed_forward_length", 11008)
        writer.add_uint32("llama.attention.head_count", 32)
        writer.add_uint32("llama.attention.head_count_kv", 32)
        writer.add_float32("llama.attention.layer_norm_rms_epsilon", 1e-5)

    # Phase 5: Tokenizer
    tok_meta = extract_tokenizer_metadata(
        kv_overrides.get("tokenizer.json"), general_kvs.get("llama.vocab_size", 0)
    )
    for k, v in tok_meta.items():
        if k not in kv_overrides:
            if isinstance(v, str):
                writer.add_string(k, v)
            elif isinstance(v, bool):
                writer.add_bool(k, v)
            elif isinstance(v, int):
                writer.add_uint32(k, v)
            elif isinstance(v, float):
                writer.add_float32(k, v)
            elif isinstance(v, list):
                if v and isinstance(v[0], str):
                    writer.add_array(k, v, GGUFValueType.STRING)
                elif v and isinstance(v[0], float):
                    writer.add_array(k, v, GGUFValueType.FLOAT32)
                elif v and isinstance(v[0], int):
                    writer.add_array(k, v, GGUFValueType.INT32)

    for k, v in kv_overrides.items():
        if not k.startswith("general."):
            if isinstance(v, bool):
                writer.add_bool(k, v)
            elif isinstance(v, str):
                writer.add_string(k, v)
            elif isinstance(v, int):
                writer.add_uint32(k, v)
            elif isinstance(v, float):
                writer.add_float32(k, v)

    # Phase 6 & 7: Tensors
    for init_name in graph.initializers:
        if init_name in graph.tensors:
            init = graph.tensors[init_name]
            if isinstance(init, Tensor):
                try:
                    gguf_name = rename_tensor(
                        init_name, kv_overrides.get("tensorNameOverrides", {})
                    )
                except ValueError:
                    continue
                shape = [int(s) if not isinstance(s, str) else 1 for s in init.shape]
                writer.add_tensor_info(gguf_name, list(reversed(shape)), get_gguf_type(init.dtype))

    writer.write_header_to_file()

    for init_name in graph.initializers:
        if init_name in graph.tensors:
            init = graph.tensors[init_name]
            if isinstance(init, Tensor) and init.data is not None:
                writer.write_tensor_data(init.data)

File: packages/js/onnx2gguf/src/compiler.ts
import { renameTensor } from './naming';
import { extractTokenizerMetadata } from './tokenizer';
import { extractMetadata, inferArchitecture } from './arch';
import { Graph, Tensor } from '@onnx9000/core';
import { GGUFWriter, GGUFValueType, GGUFTensorType } from './builder';

function getGGUFType(dtype: string): GGUFTensorType {
  if (dtype === 'float32') return GGUFTensorType.F32;
  if (dtype === 'float16') return GGUFTensorType.F16;
  return GGUFTensorType.F32;
}

function sanitizeDocString(doc: string): string {
  return doc ? doc.trim() : '';
}

export function compileGGUF(
  graph: Graph,
  kvOverrides: Record<string, any> = {},
  archOverride?: string,
): ArrayBuffer {
  const writer = new GGUFWriter();

  const arch = archOverride || inferArchitecture(graph);

  // Phase 2 Defaults
  const generalKvs: Record<string, any> = {
    'general.architecture': arch,
    'general.name': graph.name || 'model',
    'general.author': (graph as any).producerName || 'onnx9000',
    'general.version': Number((graph as any).modelVersion || 1),
    'general.quantization_version': 2,
    'general.alignment': 32,
    'general.file_type': 'mostly_f32',
  };

  const docStr = (graph as any).docString;
  if (docStr) {
    generalKvs['general.description'] = sanitizeDocString(docStr);
  }

  for (const [k, v] of Object.entries(kvOverrides)) {
    if (k.startsWith('general.')) {
      generalKvs[k] = v;
    }
  }

  for (const [k, v] of Object.entries(generalKvs)) {
    if (typeof v === 'boolean') writer.addBool(k, v);
    else if (typeof v === 'string') writer.addString(k, v);
    else if (typeof v === 'number' && !Number.isInteger(v)) writer.addFloat32(k, v);
    else if (typeof v === 'number') writer.addUint32(k, v);
  }

  // Phase 3 & 4
  const archMeta = extractMetadata(graph, archOverride);
  for (const [k, v] of Object.entries(archMeta)) {
    if (typeof v === 'boolean') writer.addBool(k, v);
    else if (typeof v === 'string') writer.addString(k, v);
    else if (typeof v === 'number' && !Number.isInteger(v)) writer.addFloat32(k, v);
    else if (typeof v === 'number') writer.addUint32(k, v);
  }

  // Phase 5: Tokenizer
  const tokMeta = extractTokenizerMetadata(
    kvOverrides['tokenizer.json'] || null,
    archMeta['llama.vocab_size'] || 0,
  );
  for (const [k, v] of Object.entries(tokMeta)) {
    if (!kvOverrides[k]) {
      if (typeof v === 'boolean') writer.addBool(k, v);
      else if (typeof v === 'string') writer.addString(k, v);
      else if (Array.isArray(v))
        writer.addArray(
          k,
          v,
          typeof v[0] === 'number'
            ? GGUFValueType.FLOAT32
            : typeof v[0] === 'string'
              ? GGUFValueType.STRING
              : /* v8 ignore start */
                GGUFValueType.INT32,
          /* v8 ignore stop */
        );
      else if (typeof v === 'number') writer.addUint32(k, v);
    }
  }

  for (const [k, v] of Object.entries(kvOverrides)) {
    if (!k.startsWith('general.') && k !== 'tokenizer.json') {
      if (typeof v === 'boolean') writer.addBool(k, v);
      else if (typeof v === 'string') writer.addString(k, v);
      else if (Array.isArray(v))
        writer.addArray(
          k,
          v,
          typeof v[0] === 'number'
            ? GGUFValueType.FLOAT32
            : typeof v[0] === 'string'
              ? GGUFValueType.STRING
              : GGUFValueType.INT32,
        );
      else if (typeof v === 'number' && !Number.isInteger(v)) writer.addFloat32(k, v);
      else if (typeof v === 'number') writer.addUint32(k, v);
    }
  }

  // Phase 6 & 7: Tensors

  let currentOffset = 0n;
  const tensorDefs: { name: string; shape: bigint[]; type: GGUFTensorType; tensor: Tensor }[] = [];

  for (const initName of graph.initializers) {
    const t = graph.tensors[initName];
    if (t) {
      const ggufName = initName.replace(/model\.layers\.(\d+)/g, 'blk.$1');
      const shape = t.shape.map((s) => (typeof s === 'number' ? BigInt(s) : 1n)).reverse();
      const type = getGGUFType(t.dtype);
      writer.addTensorInfo(ggufName, shape, type, currentOffset);
      tensorDefs.push({ name: ggufName, shape, type, tensor: t });

      let items = 1n;
      for (const d of shape) items *= d;
      let size = 0n;
      if (type === GGUFTensorType.F32) size = items * 4n;
      else if (type === GGUFTensorType.F16) size = items * 2n;

      let alignment = 32n;
      for (const kv of writer.kvs) {
        if (kv.key === 'general.alignment' && kv.type === GGUFValueType.UINT32) {
          alignment = BigInt(kv.val);
        }
      }

      const padding = (alignment - (size % alignment)) % alignment;
      currentOffset += size + padding;
    }
  }

  const headerSize = writer.getHeaderSize();
  const totalSize = BigInt(headerSize) + currentOffset;

  const buffer = new ArrayBuffer(Number(totalSize));
  const written = writer.writeHeader(buffer);
  let cursor = written;
  const u8View = new Uint8Array(buffer);

  for (const t of tensorDefs) {
    if (t.tensor.data) {
      u8View.set(
        new Uint8Array(t.tensor.data.buffer, t.tensor.data.byteOffset, t.tensor.data.byteLength),
        cursor,
      );
      let items = 1n;
      for (const d of t.shape) items *= d;
      let size = 0n;
      if (t.type === GGUFTensorType.F32) size = items * 4n;
      else if (t.type === GGUFTensorType.F16) size = items * 2n;

      let alignment = 32n;
      for (const kv of writer.kvs) {
        if (kv.key === 'general.alignment' && kv.type === GGUFValueType.UINT32) {
          alignment = BigInt(kv.val);
        }
      }

      const padding = (alignment - (size % alignment)) % alignment;
      cursor += Number(size + padding);
    }
  }

  return buffer;
}
